packages/vanilla/src/task.ts

```diff
@@ -0,0 +1,245 @@
+	constructor(name: InstanceEvents) {
+		super(name);
+	}
+}
+
+type MappedInstance<TReturn> = {
+	event_target: EventTarget;
+	is_running: boolean;
+	instance: TaskInstance<TReturn>;
+	offs: Set<() => void>;
+};
+
+export type TaskInstance<TReturn = undefined> = {
+	error?: unknown;
+	value?: TReturn;
+	on: (event: InstanceEvents, cb: () => void, options?: AddEventListenerOptions) => void;
+};
+
+function define_properties<TOn, TFrom>(on: TOn, from: TFrom): asserts on is TOn & Readonly<TFrom> {
+	for (const key in from) {
+		Object.defineProperty(on, key, {
+			enumerable: true,
+			get() {
+				return from[key as never];
+			},
+		});
+	}
+}
+
+function _task<TArgs = unknown, TReturn = undefined>(
+	gen_or_fun: TaskFunction<TArgs, TReturn>,
+	options?: TaskOptions,
+) {
+	const event_target = new EventTarget();
+
+	const task_instance = {
+		isRunning: false,
+		last: undefined as undefined | TaskInstance<TReturn>,
+		lastCanceled: undefined as undefined | TaskInstance<TReturn>,
+		lastErrored: undefined as undefined | TaskInstance<TReturn>,
+		lastRunning: undefined as undefined | TaskInstance<TReturn>,
+		lastSuccessful: undefined as undefined | TaskInstance<TReturn>,
+		performCount: 0,
+		on(event: Events, cb: () => void, options?: AddEventListenerOptions) {
+			event_target.addEventListener(event, cb, options);
+			return () => {
+				event_target.removeEventListener(event, cb, options);
+			};
+		},
+		destroy() {},
+	};
+
+	function update_is_running() {
+		task_instance.isRunning = [...instances.values()].some((val) => val.is_running);
+	}
+
+	const instances = new Map<string, MappedInstance<TReturn>>();
+
+	const actual_task = createTask<TArgs, TReturn, TaskInstance<TReturn>>(
+		{
+			onDestroy(fn) {
+				task_instance.destroy = fn;
+			},
+			onError(instance_id, error) {
+				const instance = instances.get(instance_id);
+				if (instance) {
+					task_instance.lastErrored = instance.instance;
+					task_instance.lastRunning = undefined;
+					instance.instance.error = error;
+					instance.is_running = false;
+					update_is_running();
+					instance.event_target.dispatchEvent(new SheepdogInstanceEvent('error'));
+					instance.event_target.dispatchEvent(new SheepdogInstanceEvent('finish'));
+					// we delete after a microtask to avoid returnModifier
+					// not founding the instance in case of a syncronous
+					// cancellation (for example with drop)
+					queueMicrotask(() => {
+						instances.delete(instance_id);
+						for (const off of instance.offs) {
+							off();
+						}
+					});
+				}
+				event_target.dispatchEvent(new SheepdogEvent('instance-error'));
+				event_target.dispatchEvent(new SheepdogEvent('instance-finish'));
+				if (![...instances.values()].some((instance) => instance.is_running)) {
+					event_target.dispatchEvent(new SheepdogEvent('finish'));
+				}
+			},
+			onInstanceCancel(instance_id) {
+				const instance = instances.get(instance_id);
+				if (instance) {
```

Comment:
Is the instance not being there an actual possibility or is this to please TS? If it's the former, what does that mean for the rest of the code?

(this is also true for the other `on*` handlers of course)

Reply:
I can't exactly remember but i'm pretty sure it is only to please TS...we are deleting instances on instance cancel, on error and on instance complete but in all three cases there's no way we invoke the others.

What do you mean by "the rest of the code"?